pkg/agent/openflow/client.go

```diff
@@ -470,3 +509,111 @@ func (c *client) setupPolicyOnlyFlows() error {
 	}
 	return nil
 }
+
+func (c *client) SubscribePacketIn(reason uint8, ch chan *ofctrl.PacketIn) error {
+	return c.bridge.SubscribePacketIn(reason, ch)
+}
+
+func (c *client) SendTraceflowPacket(
+	dataplaneTag uint8,
+	srcMAC string,
+	dstMAC string,
+	srcIP string,
+	dstIP string,
+	IPProtocol uint8,
+	ttl uint8,
+	IPFlags uint16,
+	TCPSrcPort uint16,
+	TCPDstPort uint16,
+	TCPFlags uint8,
+	UDPSrcPort uint16,
+	UDPDstPort uint16,
+	ICMPType uint8,
+	ICMPCode uint8,
+	ICMPID uint16,
+	ICMPSequence uint16,
+	inPort uint32,
+	outPort int32) error {
+
+	regName := fmt.Sprintf("%s%d", binding.NxmFieldReg, TraceflowReg)
+
+	packetOutBuilder := c.bridge.BuildPacketOut()
+	parsedSrcMAC, _ := net.ParseMAC(srcMAC)
+	parsedDstMAC, _ := net.ParseMAC(dstMAC)
+	if dstMAC == "" {
+		parsedDstMAC = c.nodeConfig.GatewayConfig.MAC
+	}
+
+	packetOutBuilder = packetOutBuilder.SetSrcMAC(parsedSrcMAC)
+	packetOutBuilder = packetOutBuilder.SetDstMAC(parsedDstMAC)
+	packetOutBuilder = packetOutBuilder.SetSrcIP(net.ParseIP(srcIP))
+	packetOutBuilder = packetOutBuilder.SetDstIP(net.ParseIP(dstIP))
+
+	if ttl == 0 {
+		packetOutBuilder = packetOutBuilder.SetTTL(128)
+	} else {
+		packetOutBuilder = packetOutBuilder.SetTTL(ttl)
+	}
+	packetOutBuilder = packetOutBuilder.SetIPFlags(IPFlags)
+
+	switch IPProtocol {
+	case 1:
+		packetOutBuilder = packetOutBuilder.SetIPProtocol(binding.ProtocolICMP)
+		packetOutBuilder = packetOutBuilder.SetICMPType(ICMPType)
+		packetOutBuilder = packetOutBuilder.SetICMPCode(ICMPCode)
+		packetOutBuilder = packetOutBuilder.SetICMPID(ICMPID)
+		packetOutBuilder = packetOutBuilder.SetICMPSequence(ICMPSequence)
+	case 6:
+		packetOutBuilder = packetOutBuilder.SetIPProtocol(binding.ProtocolTCP)
+		if TCPSrcPort == 0 {
+			TCPSrcPort = uint16(rand.Uint32())
+		}
+		packetOutBuilder = packetOutBuilder.SetTCPSrcPort(TCPSrcPort)
+		packetOutBuilder = packetOutBuilder.SetTCPDstPort(TCPDstPort)
+		packetOutBuilder = packetOutBuilder.SetTCPFlags(TCPFlags)
+	case 17:
+		packetOutBuilder = packetOutBuilder.SetIPProtocol(binding.ProtocolUDP)
+		packetOutBuilder = packetOutBuilder.SetUDPSrcPort(UDPSrcPort)
+		packetOutBuilder = packetOutBuilder.SetUDPDstPort(UDPDstPort)
+	}
+
+	packetOutBuilder = packetOutBuilder.SetInport(inPort)
+	if outPort != -1 {
+		packetOutBuilder = packetOutBuilder.SetOutport(uint32(outPort))
+	}
+	packetOutBuilder = packetOutBuilder.AddLoadAction(regName, uint64(dataplaneTag), OfTraceflowMarkRange)
+
+	packetOutObj := packetOutBuilder.Done()
+	return c.bridge.SendPacketOut(packetOutObj)
+}
+
+func (c *client) InstallTraceflowFlows(dataplaneTag uint8) error {
+	flow := c.traceflowL2ForwardOutputFlow(dataplaneTag, cookie.Default)
+	if err := c.Add(flow); err != nil {
+		return err
+	}
+	flow = c.traceflowConnectionTrackFlows(dataplaneTag, cookie.Default)
+	if err := c.Add(flow); err != nil {
+		return err
+	}
+	flows := []binding.Flow{}
+	for _, ctx := range c.globalConjMatchFlowCache {
```

Comment:
I think accessing `c.globalConjMatchFlowCache` need lock

Reply:
Added Lock() and defer Unlock() here.